seq2seq/optim/optim.py

```diff
@@ -1,79 +1,51 @@
 
     Args:
-        optim_class (torch.optim.Optimizer): optimizer class, e.g. torch.optim.SGD
+        optim (torch.optim.Optimizer): optimizer object, the parameters to be optimized
+            should be given when instantiating the object, e.g. torch.optim.SGD(params)
         max_grad_norm (float, optional): value used for gradient norm clipping,
             set 0 to disable (default 0)
-        lr_decay (float, optional): value for learning rate decay:
-            lr = lr_decay * lr (default 1)
-        decay_after_epoch (float, optional): learning rate starts to decay after the
-            specified epoch number, set 0 to disable (default 0)
-        **kwargs: arguments for the given optimizer class,
-            refer http://pytorch.org/docs/optim.html#algorithms for more information
     """
 
     _ARG_MAX_GRAD_NORM = 'max_grad_norm'
-    _ARG_DECAY_AFTER = "decay_after_epoch"
-    _ARG_LR_DECAY = "lr_decay"
-
-    def __init__(self, optim_class, **kwargs):
-        self.optim_class = optim_class
-        self.optimizer = None
-        self.parameters = None
-
-        self.max_grad_norm = self._get_remove(kwargs, Optimizer._ARG_MAX_GRAD_NORM, 0)
-        self.lr_decay = self._get_remove(kwargs, Optimizer._ARG_LR_DECAY, 1)
-        self.decay_after_epoch = self._get_remove(kwargs, Optimizer._ARG_DECAY_AFTER, 0)
-        self.optim_args = kwargs
 
-    def _get_remove(self, args, key, default):
-        value = default
-        if key in args:
-            value = args[key]
-            del args[key]
-        return value
-
-    def set_parameters(self, parameters):
-        """ Set the parameters to optimize.
+    def __init__(self, optim, max_grad_norm=0):
+        self.optimizer = optim
+        self.scheduler = None
+        self.max_grad_norm = max_grad_norm
 
+    def set_scheduler(self, scheduler):
+        """ Set the learning rate scheduler.
         Args:
-            parameters (iterable): An iterable of torch.nn.Parameter.
+            scheduler (torch.optim.lr_scheduler.*): object of learning rate scheduler,
+               e.g. torch.optim.lr_scheduler.StepLR
         """
-        self.parameters = parameters
-        self.optimizer = self.optim_class(parameters, **self.optim_args)
+        self.scheduler = scheduler
 
     def step(self):
         """ Performs a single optimization step, including gradient norm clipping if necessary. """
         if self.max_grad_norm > 0:
-            torch.nn.utils.clip_grad_norm(self.parameters, self.max_grad_norm)
+            params = itertools.chain.from_iterable([group['params'] for group in self.optimizer.param_groups])
+            torch.nn.utils.clip_grad_norm(params, self.max_grad_norm)
         self.optimizer.step()
 
     def update(self, loss, epoch):
-        """ Update the learning rate if the conditions are met. Override this method
-        to implement your own learning rate schedule.
-
+        """ Update the learning rate if the criteria of the scheduler are met.
```

Comment:
do you need epoch for this method ?

Reply:
No for now, but it could be useful if somebody override the method.

Reply:
👍